Replace authentication_(valid|invalid)? with (has|missing)_credentials?

jrafanie opened the pull request:
authentication_valid? is badly named since it only means we had the userid, not
that the authentication was valid.

Calling code "may" have assumed it meant that the authentication was valid.

For example, #1691 shows how client code was duped into thinking authentication_valid?
meant more than has_credentials? and caused workers to continually restart when
the authentications couldn't be validated due to bad password, unreachable systems, etc.

has_credentials? or missing_credentials? is closer to the truth.

The resulting calling code seems to read much clearer now.

Followup to #1691

Rather than keep authentication_valid? and _invalid?, it's probably better to just remove all traces of the confusing method names.

jrafanie commented on vmdb/app/models/ems_amazon.rb:
```diff
@@ -35,7 +35,7 @@ def browser_url
 
   def connect(options = {})
-    raise "no credentials defined" if self.authentication_invalid?(options[:auth_type])
+    raise "no credentials defined" if self.missing_credentials?(options[:auth_type])
```
These new lines read much easier... and actually had the same exact meaning, but now the old confusing name is gone.

Fryguy replied:
Much better :+1:

jrafanie commented on vmdb/app/models/ext_management_system.rb:
```diff
@@ -208,8 +208,7 @@ def last_refresh_status
 
   def refresh_ems
-    raise "no #{ui_lookup(:table => "ext_management_systems")} credentials defined" if self.authentication_invalid?
-    raise "refresh requires a valid user id and password" if self.authentication_invalid?
```
This second raise was not reachable.  :scream:

blomquisg replied:
:laughing:

jrafanie commented on vmdb/app/models/host.rb:
```diff
@@ -821,7 +822,7 @@ def self.refresh_ems(host_ids)
   def refresh_ems
-    raise "No #{ui_lookup(:table => "ext_management_systems")} or credentials defined" unless self.ext_management_system && self.ext_management_system.authentication_valid?
+    raise "No #{ui_lookup(:table => "ext_management_systems")} or credentials defined" unless self.ext_management_system && self.ext_management_system.has_credentials?
     EmsRefresh.queue_refresh(self)
```
Do we really want to queue a refresh if we've detected system is unavailable or we have a bad password?

blomquisg replied:
Nope!  We definitely shouldn't.

Next PR?

jrafanie replied:
No, I can fix this in an additional commit... it seems very straightforward

jrafanie commented on vmdb/app/models/host.rb:
```diff
@@ -1133,7 +1134,7 @@ def verify_credentials(auth_type=nil, options={})
 
   def verify_credentials_with_ssh(auth_type=nil, options={})
-    raise MiqException::MiqHostError, "No credentials defined" if self.authentication_invalid?(auth_type)
+    raise MiqException::MiqHostError, "No credentials defined" if self.missing_credentials?(auth_type)
```
More englishy  :smiley:

jrafanie commented on vmdb/app/models/miq_provision/options_helper.rb:
```diff
@@ -26,7 +26,7 @@ def get_source
     raise MiqException::MiqProvisionError, "#{source.class.name} [#{source.name}] is not attached to a Management System" if ems.nil?
     raise MiqException::MiqProvisionError, "#{source.class.name} [#{source.name}] is attached to <#{ems.class.name}: #{ems.name}> that does not support Provisioning" unless MiqProvision::SUPPORTED_EMS_CLASSES.include?(ems.class.name)
-    raise MiqException::MiqProvisionError, "#{source.class.name} [#{source.name}] is attached to <#{ems.class.name}: #{ems.name}> with invalid credentials" if ems.authentication_invalid?
+    raise MiqException::MiqProvisionError, "#{source.class.name} [#{source.name}] is attached to <#{ems.class.name}: #{ems.name}> with invalid credentials" if ems.missing_credentials?
```
Seems like provisioning might want to raise an error if we detect the system is unreachable or we have a bad password?  Right now, it only cares if we have the userid for the system.

blomquisg replied:
Should the exception message also say `missing credentials` instead of `invalid credentials` (for now)?

jrafanie replied:
I agree @blomquisg.

Waiting on @gmcculloug... I'm wondering if it makes sense to only check for the userid.  Bad password, unreachable provider or something else that authentication_check would detect are all things that this code won't protect against.

gmcculloug replied:
When most of these calls were implemented we did not retaining the validity of the connection to a provider and simply verified that we had proper credentials to try to connect with a remote system.  If the connection failed the caller could deal with that.  For example this code may pass this check and still fail to connect to the provider in which case the task/request message would be updated.  So the code is in fact dealing with bad password, unreachable provider and any other condition that would cause us not to connect, just not on this line.

With that being said, if we are now relying on authentication_status_ok to indicate that we should be able to connect to the provider than it makes sense to change it here.

Fryguy replied:
@jrafanie All callers still have to deal with conditions when connection might fail, because even if it says the status is ok, at the moment of connect it may not be.  So, I think we should only use authentication_status_ok as a pre-check for heavy operations...like starting the workers...don't start them unless we have known-good creds.  The worker still should expect a failure and deal with it accordingly (likely by setting the known-good status to bad).

This one, I think we should fail on known-bad connections, but the code @gmcculloug has for dealing with bad connections needs to stay (but perhaps should be updated to set the connection as bad if we *know* it's bad)

jrafanie commented on vmdb/app/models/miq_vim_broker_worker.rb:
```diff
@@ -20,7 +20,7 @@ def self.has_required_role?
 
   def self.emses_to_monitor
-    EmsVmware.where(:zone_id => MiqServer.my_server.zone_id).includes(:authentications).select(&:authentication_valid?)
+    EmsVmware.where(:zone_id => MiqServer.my_server.zone_id).includes(:authentications).select(&:has_credentials?)
```
I think this should really be calling authentication_status_ok? so we don't monitor ems's that we recently detected were unreachable, we had a bad password, etc.

blomquisg replied:
:+1:  ... Next PR?

Fryguy replied:
Yes this should be authentication_status_ok.

jrafanie commented on vmdb/app/models/mixins/authentication_mixin.rb:
```diff
@@ -34,13 +34,12 @@ def authentication_password_encrypted(type = nil)
     authentication_component(type, :password_encrypted)
   end
 
-  def authentication_valid?(type = nil)
```
:fire:

jrafanie commented on vmdb/app/models/mixins/authentication_mixin.rb:
```diff
@@ -34,13 +34,12 @@ def authentication_password_encrypted(type = nil)
   end
-  alias :has_credentials? :authentication_valid?
 
-  def authentication_invalid?(type = nil)
```
:fire:

jrafanie commented on vmdb/app/models/mixins/ems_openstack_mixin.rb:
```diff
@@ -24,7 +24,7 @@ def browser_url
     require 'openstack/openstack_handle'
     @openstack_handle ||= begin
-      raise "no credentials defined" if self.authentication_invalid?(options[:auth_type])
+      raise "no credentials defined" if self.missing_credentials?(options[:auth_type])
```
It seems like we do this this raise if missing creds all over the place ;-)

blomquisg replied:
Heh, I was thinking the same thing after the 3rd time I saw it in this PR.

kbrock replied:
A lot of this code feels like duplicates.

If we do refactor it out, would prefer adding another class (credentials) that has these methods rather than adding a mixin with a framework

Fryguy replied:
Well, there already is the Authentication class which is meant for this purpose.

blomquisg commented:
What?  A net-deletion of only 1 line?  :disappointed:

jrafanie commented on vmdb/app/models/storage.rb:
```diff
@@ -87,7 +87,7 @@ def active_hosts_with_credentials_in_zone(zone_name)
 
   def active_hosts_with_credentials
-    self.active_hosts.select(&:authentication_valid?)
+    self.active_hosts.select(&:has_credentials?)
```
This is called when we want to scan the storage... seems like another place to use authentication_status_ok? instead.

Fryguy replied:
Yes this should be authentication_status_ok.  @gmcculloug ?

gmcculloug replied:
Agreed, should be authentication_status_ok now.

jrafanie commented on vmdb/app/models/vm_or_template.rb:
```diff
@@ -1064,7 +1064,7 @@ def storage2active_proxies(all_proxy_list = nil)
     # Remove any MiqServer instances if we do not have credentials
     rsc = self.scan_via_ems? ? self.ext_management_system : self.host
-    proxies.delete_if {|p| MiqServer === p} if rsc && rsc.authentication_invalid?
+    proxies.delete_if {|p| MiqServer === p} if rsc && rsc.missing_credentials?
```
Again, when scanning(fleecing) a vm, do we really want to try a MiqProxy that is unreachable, has a bad password?

Fryguy replied:
Not sure on this one @gmcculloug ?

jrafanie replied:
Squirrel/shiny object:  Note, it's really a shame we can't bubble up that these "proxies" were eliminated due to missing creds, bad authentication, bad response (down)... the user will end up with "no eligible proxies" and have no idea why...

gmcculloug replied:
Use authentication_status_ok

Fryguy replied:
Aside...I thought this proxy stuff was removed...or is there still remnants needed?

gmcculloug replied:
This is MiqServer embedded proxy logic which does not change with the removal of the external SmartProxy.

Fryguy replied:
Ok, so the concept of embedded proxy still stands...I thought that was thrown away too (or at least the concept that it's a proxy as opposed to just being the code that is called)

gmcculloug replied:
@jrafanie As luck would have it we do have logic in place to notify users that proxies were eliminated.  The code could use some updates but the mechanism is in place. https://github.com/ManageIQ/manageiq/blob/master/vmdb/app/models/vm_or_template.rb#L980

jrafanie commented on vmdb/app/models/vm_or_template.rb:
```diff
@@ -1165,7 +1165,7 @@ def self.refresh_ems(vm_ids)
 
   def refresh_ems
-    raise "no #{ui_lookup(:table => "ext_management_systems")} or credentials defined" unless self.ext_management_system && self.ext_management_system.authentication_valid?
+    raise "no #{ui_lookup(:table => "ext_management_systems")} or credentials defined" unless self.ext_management_system && self.ext_management_system.has_credentials?
```
Another place to use authentication_status_ok?  I don't want to refresh if the system not reachable or our password is bad.

Fryguy replied:
Same answer as my question above...I think we should check both (or does authentication_status_ok handle missing credentials as well?)

jrafanie replied:
authentication_check is responsible for populating the status with incomplete if we don't have credentials.  See here: https://github.com/jrafanie/manageiq/blob/82027846ff1aab1943cad86c97ffc1ff9c065513/vmdb/app/models/mixins/authentication_mixin.rb#L163

It should detect many situations such as no userid, bad password, unreachable, some other error:
https://github.com/jrafanie/manageiq/blob/82027846ff1aab1943cad86c97ffc1ff9c065513/vmdb/app/models/authentication.rb#L28

jrafanie commented on vmdb/app/models/vm_or_template.rb:
```diff
@@ -1176,12 +1176,12 @@ def self.refresh_ems_sync(vm_ids)
 
   def refresh_on_reconfig
-    raise "no #{ui_lookup(:table => "ext_management_systems")} or credentials defined" unless self.ext_management_system && self.ext_management_system.authentication_valid?
+    raise "no #{ui_lookup(:table => "ext_management_systems")} or credentials defined" unless self.ext_management_system && self.ext_management_system.has_credentials?
```
Same as above for the above two.

jrafanie commented on vmdb/lib/workers/vim_broker_worker.rb:
```diff
@@ -25,7 +25,7 @@ def after_initialize
     emses = MiqVimBrokerWorker.emses_to_monitor
     MiqPreloader.preload(emses, :hosts => :authentications)
-    hosts = emses.collect(&:hosts).flatten.uniq.select(&:authentication_valid?)
+    hosts = emses.collect(&:hosts).flatten.uniq.select(&:has_credentials?)
```
Again, authentication_status_ok?...

Fryguy replied:
Yes.

jrafanie commented:
cc @blomquisg @Fryguy @brandondunne @gmcculloug @lfu @roliveri @tenderlove @chessbyte @kbrock @lfu @bzwei  If you've ever used authentication_valid? or authentication_invalid? on our Ems/Hosts/FileDepots, please see above... I'm removing it because it's confusing.  

Use this handy cheat sheet for when to use the remaining methods:
has_credentials?  - it has credentials
missing_credentials?  - it is missing credentials
authentication_status_ok? - the last authentication check against the remote system was "Valid"

jrafanie commented:
Yes, only 1 net deletion... :cry:

blomquisg commented on vmdb/app/models/host.rb:
```diff
@@ -789,8 +789,9 @@ def refreshable_status
-        if self.authentication_invalid?(:ipmi)
+      # TODO: Simplify these conditions
+      if self.has_credentials?(:ipmi) || !self.ipmi_address.blank?
+        if self.missing_credentials?(:ipmi)
```
:+1: That **is** a confusing set of conditions...

Maybe this?
```ruby
s.merge!( ..., :message => "Provide an IPMI Address") if self.has_credentials?(:ipmi) && self.ipmi_address.blank?
s.merge!( ..., :message => "Provide credentials for IPMI") if self.missing_credentials?(:ipmi) && !self.ipmi_address.blank?
```

Fryguy replied:
:confused:

jrafanie replied:
Yeah, I need a truth table.  We need to put this under test and see what bugs shake out.  At the very least, it's hard to comprehend why we do

```if self.authentication_valid?(:ipmi) || !self.ipmi_address.blank?```

Then do
```if self.ipmi_address.blank?```

Fryguy replied:
| has impi creds | ipmi address set | then do |
| --- | --- | --- |
| true | true | "" |
| true | false | Provide an IPMI Address |
| false | true | Provide credentials for IPMI |
| false | false | falls through, so ???? |

jrafanie replied:
yeah, the false/false case is strange.

Fryguy replied:
I think false/false should just be the same as true/false.  Basically, you didn't provide either, so at least prompt for one.

blomquisg commented:
:100: This whole thing reads much better than before!  Good job

Fryguy commented:
@jrafanie I think we should fix the "wrong" usages of has_credentials in this PR, just in follow up commits, based on input from all the people mentioned above.  I'll comment on what I know.

Also, `authentication_status_ok?` depends on something coming along and actually verifying the credentials.  For EMSes, I believe this is scheduled.  However, is there something in place for other authentications, like, say, FileDepot?  Even so, I believe that when someone clicks verify in the UI (i.e. when verify_credentials is called), then we should set the authentication_status to ok at that time, since at least at that moment we are sure of it.

Fryguy commented on vmdb/app/models/ext_management_system.rb:
```diff
@@ -208,8 +208,7 @@ def last_refresh_status
   def refresh_ems
-    raise "no #{ui_lookup(:table => "ext_management_systems")} credentials defined" if self.authentication_invalid?
-    raise "refresh requires a valid user id and password" if self.authentication_invalid?
+    raise "no #{ui_lookup(:table => "ext_management_systems")} credentials defined" if self.missing_credentials?
```
Wondering if there should be a second raise here when `!self.authentication_status_ok`.  @blomquisg Thoughts?

blomquisg replied:
Yeah, I agree.  Fail on missing credentials, and also fail on bad authentication status.

Fryguy commented on vmdb/app/models/mixins/file_depot_mixin.rb:
```diff
@@ -53,7 +53,7 @@ def requires_credentials?
 
   def validate_depot_credentials
     # This only checks that credentials are present
```
At least the code here actually *said* what it expected :smile:   cc @brandondunne

blomquisg replied:
HA!

jrafanie replied:
There's no fooling @brandondunne, he's a sharp chap.

Fryguy commented on vmdb/app/models/vm_or_template.rb:
```diff
@@ -328,7 +328,7 @@ def refresh_state
   def run_command_via_parent(verb, options = {})
     # TODO: Need to break this logic out into a method that can look at the verb and the vm and decide the best way to invoke it - Virtual Center WS, ESX WS, Storage Proxy.
-    if self.ext_management_system && self.ext_management_system.authentication_valid? && self.ext_management_system.respond_to?(verb)
+    if self.ext_management_system && self.ext_management_system.has_credentials? && self.ext_management_system.respond_to?(verb)
```
This method is for actually running commands on the provider...I *think* it should be authentication_status_ok.  @gmcculloug ?

P.S. I hate this method :smile_cat:

gmcculloug replied:
Use authentication_status_ok

Fryguy commented on vmdb/spec/models/mixins/authentication_mixin_spec.rb:
```diff
@@ -258,17 +258,17 @@ class TestClass < ActiveRecord::Base
+        @ems.has_credentials?.should be_true
       end
 
       it "should have invalid authentication if userid nil" do
```
I think these it's should have their text updated to be as clear as the new method names.

jrafanie replied:
Good point

Fryguy commented:
@jrafanie I comment on all the parts I recognized.  Additionally, I think all of the `.connect` methods and `.verify_credentials` methods are correct as is with the `missing_credentials?` check, and should not change to `authentication_status_ok`

blomquisg commented:
> Additionally, I think all of the `.connect` methods and `.verify_credentials` methods are correct as is with the `missing_credentials?` check, and should not change to `authentication_status_ok`

K, this makes sense.  But, we should be using `authentication_status_ok` in places where we're about to make connections, like before refreshes and scans.

Fryguy commented:
> where we're about to make connections, like before refreshes and scans.

Yes, as long as there is a mechanism outside of that code that checks the credentials on some regular basis.  Don't want to get into a situation where it temporarily goes bad, and then it can never connect again.

Incidentally, we probably should immediately mark it as "bad" if during the refresh or scan we detect a bad set of credentials.

jrafanie commented:
> Also, authentication_status_ok? depends on something coming along and actually verifying the credentials. For EMSes, I believe this is scheduled. However, is there something in place for other authentications, like, say, FileDepot? Even so, I believe that when someone clicks verify in the UI (i.e. when verify_credentials is called), then we should set the authentication_status to ok at that time, since at least at that moment we are sure of it.

So, I believe we only check if the authentication [was changed](https://github.com/jrafanie/manageiq/blob/82027846ff1aab1943cad86c97ffc1ff9c065513/vmdb/app/models/authentication.rb#L85).  If it was previously "bad" because the system was down and we click validate, it wouldn't change "status".

When you click validate in the UI, it could be using credentials that are not yet persisted since it's very likely someone could be entering a Ems/Host/etc for the first time and wants to validate the creds before saving.    We'd have to change the UI code to call Authentication#authentication_check(type) instead of verify_credentials if we have an authentication.

jrafanie commented:
> Yes, as long as there is a mechanism outside of that code that checks the credentials on some regular basis. Don't want to get into a situation where it temporarily goes bad, and then it can never connect again.

Yes, agreed.  Maybe the schedule of once per day is a bit paranoid.  
 
> Incidentally, we probably should immediately mark it as "bad" if during the refresh or scan we detect a bad set of credentials.

If we use authentication_check(type) instead of verify_credentials wherever the latter is used, it will mark the creds as bad and return true/false if it was successful.

jrafanie commented on vmdb/app/models/host.rb:
```diff
@@ -788,18 +788,18 @@ def refreshable_status
     end
-    return s
+
+    s
```
@Fryguy @blomquisg does this look a bit better?  I still have to write tests for this but wanted to get started on this.

jrafanie commented on vmdb/app/models/host.rb:
```diff
@@ -788,18 +788,18 @@ def refreshable_status
+    elsif self.ipmi_address.blank?
+      s.merge!(:show => true, :enabled => false, :message => "Provide an IPMI Address")
+    elsif self.missing_credentials?(:ipmi)
+      s.merge!(:show => true, :enabled => false, :message => "Provide credentials for IPMI")
```
Sorry, commented on the wrong git hunk.  @Fryguy @blomquisg does this look a bit better? I still have to write tests for this but wanted to get started on this.

jrafanie commented:
So, @Fryguy @blomquisg I have applied all of the suggested changes.  I still have to clean up the last commit to add some tests but I think I got everything else.  Did I miss anything?  See the last 6 commits.

jrafanie commented:
@miq-bot add_label needs changelog

jrafanie commented:
Wow, @Fryguy @blomquisg @gmcculloug @kbrock, this is finally passing and I *think* I addressed all of your comments without chasing too many :sparkles: objects

Fryguy commented:
Looks good to me! :+1: @blomquisg @gmcculloug If you are good with this, please merge.

jrafanie commented:
@blomquisg @gmcculloug please review/merge :bow: :heart:

gmcculloug commented:
Nice job @jrafanie  :star:

jrafanie commented:
:tada: Yay!!!

kbrock commented:
:cake: yay
